docs/build/voting-dapp-tutorial.md

```diff
@@ -0,0 +1,428 @@
+
+## Environment Setup
+
+### Install Pact
+
+**Using Brew**
+
+```clojure
+brew update
+brew install kadena-io/pact/pact
+```
+**Build from source**
+- follow the instructions [here](https://github.com/kadena-io/pact#building-from-source)
+
+**Verifying Installation**
+
+Type `pact` in a terminal and try out some commands:
+
+```clojure
+$ pact
+pact> (+ 1 2)
+3
+pact> (+ "hello, " "world")
+"hello, world"
+pact> (< 0 1)
+true
+```
+
+### Create Project Structure
+
+Copy paste the instructions below to create a basic project structure:
+
+```clojure
+mkdir my-pact-project && cd my-pact-project
+mkdir pact
+mkdir app
+```
+
+### Pact Http Server
+
+Create a `config.yaml` file for pact http server. We will later use it simulate blockchains transactions locally.
+
+For more details about pact http server check <a href="/blog/deploy-to-a-local-server">this</a> tutorial.
+
+```clojure
+touch config.yaml
+```
+Copy the code below and save the file:
+
+```clojure
+# Config file for pact http server. Launch with `pact -s config.yaml`
+
+# HTTP server port
+port: 8080
+
+# directory for HTTP logs
+logDir: log
+
+# persistence directory
+persistDir: log
+
+# SQLite pragmas for pact back-end
+pragmas: []
+
+# verbose: provide log output
+verbose: True
+```
+
+### Atom IDE (Optional)
+
+You can use any text editor to write Pact but if you prefer the benefits of an IDE, "language-pact" is a package for [Atom](https://atom.io) that provides syntax highlighting and linting.
+
+To install it, go to Preferences -> Packages -> Type "language-pact" and click the Install button.
+
+![atom-language-pact](/img/docs/voting-dapp/atom-language-pact.png)
+
+## Smart Contracts
+
+Our voting app should allow you to submit a vote while not allowing an address to vote more than once.
+Additionally we will use a *gas station* to fund the gas fees for interacting with our contract, meaning our users don't have to worry about paying gas.
+
+### Voting
+
+Let's start to implement our voting contract.
+
+Assuming you are still in your project directory, create the `vote.pact` and `vote.repl` files:
+
+```clojure
+touch pact/vote.pact pact/vote.repl
+```
+
+`vote.pact` is the contract source code while in `vote.repl` we'll write tests.
+
+Copy the following code in the `vote.pact` file:
+
+```clojure
+(define-keyset 'vote-admin-keyset)
+
+(module simple-vote 'vote-admin-keyset
+  "Simple voting module"
+
+  (defun vote (key)
+    "Submit a new vote"
+    (format "Voted {}!" [key]))
+)
+
+(vote "A")
+```
+
+This snippet defines a module "simple-vote" that holds a function "vote" that takes a parameter and displays a formatted string that includes the parameter.
+
+Before we move forward, let's quickly test our code. Copy the code below in the `vote.repl` file:
+
+```clojure
+;; begin a transaction
+(begin-tx)
+;; set transaction signature key to my-key
+(env-keys ["my-key"])
```

Comment:
Explaiin the notion of "repl-only" functions, and explain that these functions are simulating the api structures that are represented by the chainweb-blockchain

Reply:
Explained it.